nixos/modules/services/networking/webhook.nix

```diff
@@ -0,0 +1,163 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.webhook;
+  defaultUser = "webhook";
+
+  hooksFile = pkgs.writeText "webhook-config" cfg.hooks;
+
+in {
+  options = {
+    services.webhook = {
+      enable = mkEnableOption (mdDoc ''
+        [Webhook]. It's a server written in Go that allows you to create HTTP endpoints (hooks),
+        which execute configured commands for any person or service that knows the URL.
+
+        [Webhook]: https://github.com/adnanh/webhook
+      '');
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.webhook;
+        defaultText = literalExpression "pkgs.webhook";
+        example = literalExpression "pkgs.webhook";
+        description = mdDoc ''
+          Which Webhook package to use
+
+          This allows passing different version of the package.
+        '';
+      };
+      user = mkOption {
+        type = types.str;
+        default = defaultUser;
+        description = mdDoc ''
+          Webhook will be run under this user.
+
+          If set, you must create this user yourself!
+        '';
+      };
+      group = mkOption {
+        type = types.str;
+        default = defaultUser;
+        description = mdDoc ''
+          Webhook will be run under this group.
+
+          If set, you must create this group yourself!
+        '';
+      };
+      ip = mkOption {
+        type = types.str;
+        default = "0.0.0.0";
+        description = mdDoc ''
+          The IP webhook should serve hooks on.
+
+          The default means it be reached on any interface if `openFirewall = true`.
+          Otherwise, it can only be reached from the localhost.
+        '';
+      };
+      port = mkOption {
+        type = types.port;
+        default = 9000;
+        description = mdDoc "The port webhook should be reachable from.";
+      };
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          Open the configured port in the firewall for external ingress traffic.
+          Preferably the Webhook server is instead put behind a reverse proxy.
+        '';
+      };
+      enableTemplate = mkOption {
+        type = types.bool;
+        default = false;
+        description = mdDoc "Enable the generated hooks file to be parsed as a Go template.";
+      };
+      urlPrefix = mkOption {
+        type = types.str;
+        default = "hooks";
+        example = "protocol://yourserver:port/PREFIX/:hook-id";
+        description = mdDoc ''
+          The URL path prefix to use for served hooks.
+        '';
+      };
+      hooks = mkOption {
+        type = types.str;
```

Comment:
I'm thinking this option's type should probably be JSON instead of an opaque string, or maybe something even more specific like

```nix
let hooksFormat = pkgs.formats.json {}; in
types.attrsOf (types.submodule ({ name, ... }: {
  freeformType = hooksFormat.type;
  options = {
    id = {
      type = types.str;
      default = name;
    };
    execute-command = {
      type = types.str;
    };
  };
}))
```

...and then use `hooksFormat.generate "webhook-config.json" (builtins.attrValues cfg.hooks)` or something like that. This makes it easier to define hooks across multiple modules, and we can also add an assertion checking that at least one hook is defined.

See https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md for more information.

I think this should be done in this PR so that we get the interface right the first time and don't have to do painful migrations in the future. Don't hesitate to ask if you need help implementing this.

One thing to ponder is how this interacts with [`-template`](https://github.com/adnanh/webhook/blob/master/docs/Templates.md). It seems like in typical usage, the template syntax would be confined inside JSON strings, so there should no problem translating it to Nix syntax. If more flexibility turns out to be needed, we can always add a way to configure a hook using a single string in the future.

Reply:
> the template syntax would be confined inside JSON strings, so there should no problem translating it to Nix syntax

I guess there are problems with escaping though, e.g.
```
"secret": "{{ getenv "XXXTEST_SECRET" | js }}",
```
isn't valid JSON, so we can't generate that with `pkgs.formats.json`.

It seems like webhook supports passing multiple `-hooks` flags, so we could have one for each hook, and each hook could be either a structured submodule or a literal string.

```nix
hooks.hello = { execute-command = "hello"; };
hooks.helloTemplate = ''{"id": "hello-template", "execute-command": "{{hello "template" | whatever}}"}'';
```

Reply:
This was modified in favor of a string in 3c65372ac93a61ba1a38c9920869f3d92b6aedf2 as per [this comment](https://github.com/NixOS/nixpkgs/pull/205636#discussion_r1051495119) in the current PR.

Thank you very much for the comments. I learned more about options and discovered that probably a few things could be automated e.g checking that `enable` and `package` are of certain option types. 
To be honest though, I'm kind of mentally done with this PR and am going to move on to other things I was planning on working on and learning during the holiday season. Reading nixos RFCs weren't on my list :wink:

Reply:
Oh, then it seems like we can start with reverting that commit and adding a `freeformType` as suggested.

> To be honest though, I'm kind of mentally done with this PR

That's fine, I can take over. Thanks for the work done!

Reply:
> That's fine, I can take over. Thanks for the work done!

![image](https://user-images.githubusercontent.com/2829538/209815806-3bfe89dd-8405-41c0-8479-ae8169970aa5.png)

You have edit access on the branch. Thanks for taking over